src/metabase/feature_extraction/automagic_dashboards.clj

```diff
@@ -0,0 +1,335 @@
+
+(defn- create-card!
+  [{:keys [visualization title description query]}]
+  (let [[visualization visualization-settings] (if (sequential? visualization)
+                                                 visualization
+                                                 [visualization {}])
+        card (db/insert! Card
+               :creator_id             api/*current-user-id*
+               :dataset_query          query
+               :description            description
+               :display                (or visualization :table)
+               :name                   title
+               :visualization_settings visualization-settings
+
+               :result_metadata        (card.api/result-metadata-for-query query)
+               :collection_id          (-> automagic-collection deref :id))]
+    (events/publish-event! :card-create card)
+    (hydrate card :creator :dashboard_count :labels :can_write :collection)
+    card))
+
+(defn- add-to-dashboard!
+  [dashboard card]
+  (dashboard/add-dashcard! dashboard (create-card! card)
+    (merge (next-card-position dashboard)
+           {:sizeX card-width
+            :sizeY card-height})))
+
+(defn- ensure-seq
+  [x]
+  (if (sequential? x)
+    x
+    [x]))
+
+(defn- build-query
+  [database table-id filters metrics dimensions limit order_by]
+  (let [query {:type     :query
+               :database database
+               :query    (cond-> {:source_table table-id}
+                           (not-empty filters)
+                           (assoc :filter (apply vector :and filters))
+
+                           (not-empty dimensions)
+                           (assoc :breakout dimensions)
+
+                           metrics
+                           (assoc :aggregation metrics)
+
+                           limit
+                           (assoc :limit limit))}]
+    (when (perms/set-has-full-permissions-for-set?
+           @api/*current-user-permissions-set*
+           (card/query-perms-set query :write))
+      query)))
+
+(defn- card-candidates
+  [context {:keys [metrics filters dimensions score limit order_by] :as card}]
+  (let [filters    (some->> filters
+                            ensure-seq
+                            (map (partial get (:filters context))))
+        dimensions (some->> dimensions
+                            ensure-seq
+                            (map (partial get (:dimensions context))))
+        metrics    (some->> metrics
+                            ensure-seq
+                            (map (partial get (:metrics context))))
+        bindings   (concat filters dimensions metrics)]
+    (when (every? some? bindings)
+      (let [score (* (or score max-score)
+                     (/ (transduce (map :score) + bindings)
+                        max-score
+                        (count bindings)))]
+        (->> (combo/cartesian-product
+              (apply combo/cartesian-product (map :matches filters))
```

Comment:
putting all of this into a single function is needlessly complicated and prevents proper error handling. We should break up the process of specifying filters, dimensions and metrics into separate functions and apply error handling there.